### PR TITLE
SCUMM HE: Use Miles AdLib driver

### PR DESCRIPTION
**Pull request by nukeykt**
This patch replaces generic AdLib driver with Miles AdLib driver for Humongous games. This resolves 7880 and 6788 tickets.

### CODE REVIEW

**Comment by wjp** [comment]
Thank you for the contribution!

Could you give a more in-depth explanation? Both the pull request and the commit message are lacking any information. (Also, there seem to be a few unrelated changes in there?)

**Comment by nukeykt** [comment]
Ok.
I've added special AdLib player. This player extracts FM timbres from DRV resource file and opens Miles driver.
I've also made some changes to music APIs. Humongous games pass two parameters when starting song: sound id and track id.
Also i've added master volume control and fixed pitchbend bug in Miles driver.
The final change is Putt Putt Goes to the Moon Demo tweak. PuttMoon unlike other games has GM music, so it uses different timbre file, but its demo version reuses Putt Putt Parade's music. To resolve this I've just changed demo's game id to generic HEGAME.

File: audio/miles_adlib.cpp [hunk 1]
```diff
@@ -723,7 +726,7 @@ void MidiDriver_Miles_AdLib::updatePhysicalFmVoice(byte virtualFmVoice, bool key
 
 	if (registerUpdateFlags & kMilesAdLibUpdateFlags_Reg_40) {
 		// Calculate new volume
-		byte midiVolume = _midiChannels[midiChannel].currentVolume;
+		byte midiVolume = (_midiChannels[midiChannel].currentVolume * _masterVolume) >> 8;
```
Comment (sev-) [inline audio/miles_adlib.cpp]: Why are you changing here the behaviour of the driver? The rest of the driver users are not aware of this change, and still pass volume as a byte, and will effectively mute sound now.

File: engines/scumm/players/player_he.cpp [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+		drvName = _vm->_filenamePattern.pattern;
+		drvName += '.';
+	}
+
+	drvName += "drv";
+
+	if (!file.open(drvName))
+		error("Player_HE::loadAdLibBank(): could not open %s", drvName.c_str());
+
+	uint32 size = uint32(file.size());
```
Comment (sev-) [inline engines/scumm/players/player_he.cpp]: Is it a type casting? Please use canonical way: (uint32)file.size();

File: engines/scumm/players/player_he.cpp [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+
```
Comment (sev-) [inline engines/scumm/players/player_he.cpp]: This line has trailing whitespaces

File: audio/mididrv.h [hunk 4]
```diff
@@ -213,6 +213,9 @@ class MidiDriver : public MidiDriver_BASE {
 	/** Close the midi driver. */
 	virtual void close() = 0;
 
+	/** Set global music volume */
+	virtual void setVolume(int volume) { }
```
Comment (sev-) [inline audio/mididrv.h]: This is hides overloaded functions at engines/sci/sound/drivers/amigamac.cpp, engines/parallaction/adlib.cpp, engines/agos/drivers/accolade/adlib.cpp, /engines/queen/midiadlib.h, engines/sherlock/scalpel/drivers/adlib.cpp which define it as setVolume(byte)

File: engines/scumm/music.h [hunk 5]
```diff
@@ -55,6 +55,13 @@ class MusicEngine {
 	 */
 	virtual void startSound(int sound) = 0;
 
+	/**
+	 * Start playing the sound with the given id and track id.
+	 * @param sound		the sound to start
+	 * @param track		the track to start
+	 */
+	virtual void startSound(int sound, int track) { startSound(sound); }
```
Comment (sev-) [inline engines/scumm/music.h]: This shadows overloaded startSound() in engines/scumm/imuse_digi/dimuse.h.

**Comment by sev-** [comment]
I forgot to mention that the merge commit must be not part of this PR too. In order to avoid it, just rebase your patch.

File: audio/miles_adlib.cpp (hunk 1 above)
Comment (nukeykt) [reply audio/miles_adlib.cpp]: AFAIK all other games handle volume control outside of this module. We defaulting this value to 255, so other games should work fine.

File: audio/miles_adlib.cpp [hunk 6]
```diff
@@ -270,7 +270,7 @@ class MidiDriver_Miles_AdLib : public MidiDriver {
 };
 
 MidiDriver_Miles_AdLib::MidiDriver_Miles_AdLib(InstrumentEntry *instrumentTablePtr, uint16 instrumentTableCount)
-	: _masterVolume(15), _opl(0),
+	: _masterVolume(255), _opl(0),
```
Comment (sev-) [inline audio/miles_adlib.cpp]: Were other engines which use this driver tested with this change? Why this change is required at all?

**Comment by sev-** [comment]
Very good. Thank you a lot for the change. Merging now.

Also, if you're interested in working further on HE sounds, I have something to discuss on the IRC. Come, talk to me (nick _sev at #scummvm on freenode)

**Comment by nukeykt** [comment]
I am busy with other projects now, but i'm still interested. I'll contact you when i have a spare time.
